File: scripts/generate-synthetic-data.ts
import fs from "fs";
import path from "path";

// Deterministic PRNG (Mulberry32)
function mulberry32(a: number) {
  return function() {
    let t = a += 0x6D2B79F5;
    t = Math.imul(t ^ t >>> 15, t | 1);
    t ^= t + Math.imul(t ^ t >>> 7, t | 61);
    return ((t ^ t >>> 14) >>> 0) / 4294967296;
  }
}

// Fixed seed
const seed = 123456789;
const random = mulberry32(seed);

function generate() {
  const baseDate = new Date("2026-08-18");
  
  const payments: any[] = [];
  const settlements: any[] = [];
  const bank: any[] = [];
  const ledgers: any[] = [];
  const groundTruth: any[] = [];

  let nextId = 1;

  function addMatch(source: string, target: string, expectedMatch: boolean = true) {
    groundTruth.push({ source_transaction: source, target_transaction: target, expected_match: expectedMatch.toString() });
  }

  // 1. Exact matches (40)
  for (let i = 0; i < 40; i++) {
    const id = nextId++;
    const ref = `PAY-${1000 + id}`;
    const amount = (random() * 10000 + 100).toFixed(2);
    const dateStr = baseDate.toISOString().split("T")[0];
    
    payments.push({
      externalId: `P-${id}`, amount, currency: "INR", transactionDate: dateStr, referenceId: ref, merchantName: "Acme Ltd"
    });
    settlements.push({
      externalId: `S-${id}`, amount, currency: "INR", transactionDate: dateStr, referenceId: ref, description: "ACME LTD SETTLEMENT"
    });
    bank.push({
      externalId: `B-${id}`, amount, currency: "INR", transactionDate: dateStr, referenceId: ref, description: "ACH SETTLEMENT ACME"
    });
    ledgers.push({
      externalId: `L-${id}`, amount, currency: "INR", transactionDate: dateStr, referenceId: `INV-${900 + id}`, customerName: "Acme Limited"
    });

    addMatch(`P-${id}`, `S-${id}`);
    addMatch(`S-${id}`, `B-${id}`);
    addMatch(`B-${id}`, `L-${id}`);
  }

  // 2. Date offsets (T+1, T+2) (15)
  for (let i = 0; i < 15; i++) {
    const id = nextId++;
    const ref = `PAY-${1000 + id}`;
    const amount = (random() * 10000 + 100).toFixed(2);
    const pDate = new Date(baseDate);
    const sDate = new Date(baseDate); sDate.setDate(sDate.getDate() + 1); // T+1
    const bDate = new Date(baseDate); bDate.setDate(bDate.getDate() + 2); // T+2
    const lDate = new Date(baseDate);

    payments.push({
      externalId: `P-${id}`, amount, currency: "INR", transactionDate: pDate.toISOString().split("T")[0], referenceId: ref, merchantName: "Acme Ltd"
    });
    settlements.push({
      externalId: `S-${id}`, amount, currency: "INR", transactionDate: sDate.toISOString().split("T")[0], referenceId: ref, description: "ACME LTD SETTLEMENT"
    });
    bank.push({
      externalId: `B-${id}`, amount, currency: "INR", transactionDate: bDate.toISOString().split("T")[0], referenceId: ref, description: "ACH SETTLEMENT ACME"
    });
    ledgers.push({
      externalId: `L-${id}`, amount, currency: "INR", transactionDate: lDate.toISOString().split("T")[0], referenceId: `INV-${900 + id}`, customerName: "Acme Limited"
    });

    addMatch(`P-${id}`, `S-${id}`);
    addMatch(`S-${id}`, `B-${id}`);
    addMatch(`B-${id}`, `L-${id}`);
  }

  // 3. Amount mismatches (fees, entry errors) (10)
  for (let i = 0; i < 10; i++) {
    const id = nextId++;
    const ref = `PAY-${1000 + id}`;
    const baseAmount = random() * 10000 + 100;
    const dateStr = baseDate.toISOString().split("T")[0];

    payments.push({
      externalId: `P-${id}`, amount: baseAmount.toFixed(2), currency: "INR", transactionDate: dateStr, referenceId: ref, merchantName: "Acme Ltd"
    });
    settlements.push({
      externalId: `S-${id}`, amount: (baseAmount - 1.5).toFixed(2), currency: "INR", transactionDate: dateStr, referenceId: ref, description: "ACME LTD SETTLEMENT (NET)" // Fee deducted
    });
    bank.push({
      externalId: `B-${id}`, amount: (baseAmount - 1.5).toFixed(2), currency: "INR", transactionDate: dateStr, referenceId: ref, description: "ACH SETTLEMENT ACME"
    });
    ledgers.push({
      externalId: `L-${id}`, amount: baseAmount.toFixed(2), currency: "INR", transactionDate: dateStr, referenceId: `INV-${900 + id}`, customerName: "Acme Limited"
    });

    addMatch(`P-${id}`, `S-${id}`);
    addMatch(`S-${id}`, `B-${id}`);
    addMatch(`B-${id}`, `L-${id}`);
  }

  // 4. Missing references (10)
  for (let i = 0; i < 10; i++) {
    const id = nextId++;
    const ref = `PAY-${1000 + id}`;
    const amount = (random() * 10000 + 100).toFixed(2);
    const dateStr = baseDate.toISOString().split("T")[0];

    payments.push({
      externalId: `P-${id}`, amount, currency: "INR", transactionDate: dateStr, referenceId: ref, merchantName: "Acme Ltd"
    });
    settlements.push({
      externalId: `S-${id}`, amount, currency: "INR", transactionDate: dateStr, referenceId: ref, description: "ACME LTD SETTLEMENT"
    });
    bank.push({
      externalId: `B-${id}`, amount, currency: "INR", transactionDate: dateStr, referenceId: "", description: "ACH SETTLEMENT ACME" // Missing ref
    });
    ledgers.push({
      externalId: `L-${id}`, amount, currency: "INR", transactionDate: dateStr, referenceId: `INV-${900 + id}`, customerName: "Acme Limited"
    });

    addMatch(`P-${id}`, `S-${id}`);
    addMatch(`S-${id}`, `B-${id}`);
    addMatch(`B-${id}`, `L-${id}`);
  }

  // 5. Description variations (10)
  for (let i = 0; i < 10; i++) {
    const id = nextId++;
    const ref = `PAY-${1000 + id}`;
    const amount = (random() * 10000 + 100).toFixed(2);
    const dateStr = baseDate.toISOString().split("T")[0];

    payments.push({
      externalId: `P-${id}`, amount, currency: "INR", transactionDate: dateStr, referenceId: ref, merchantName: "Stripe Inc"
    });
    settlements.push({
      externalId: `S-${id}`, amount, currency: "INR", transactionDate: dateStr, referenceId: ref, description: "STRIPE PAYMENT SETTLEMENT"
    });
    bank.push({
      externalId: `B-${id}`, amount, currency: "INR", transactionDate: dateStr, referenceId: ref, description: "MISC DEPOSIT STRIPE.COM 88392"
    });
    ledgers.push({
      externalId: `L-${id}`, amount, currency: "INR", transactionDate: dateStr, referenceId: `INV-${900 + id}`, customerName: "Stripe Corporation"
    });

    addMatch(`P-${id}`, `S-${id}`);
    addMatch(`S-${id}`, `B-${id}`);
    addMatch(`B-${id}`, `L-${id}`);
  }

  // 6. Duplicate-like records (10)
  for (let i = 0; i < 5; i++) { // 5 pairs = 10 records
    const id1 = nextId++;
    const id2 = nextId++;
    const amount = (random() * 10000 + 100).toFixed(2);
    const dateStr = baseDate.toISOString().split("T")[0];

    // True duplicate in payments (same ref, same amount)
    payments.push({
      externalId: `P-${id1}`, amount, currency: "INR", transactionDate: dateStr, referenceId: `PAY-DUP-${i}`, merchantName: "Acme Ltd"
    });
    payments.push({
      externalId: `P-${id2}`, amount, currency: "INR", transactionDate: dateStr, referenceId: `PAY-DUP-${i}`, merchantName: "Acme Ltd"
    });
    
    // Only one settlement and bank record
    settlements.push({
      externalId: `S-${id1}`, amount, currency: "INR", transactionDate: dateStr, referenceId: `PAY-DUP-${i}`, description: "ACME LTD SETTLEMENT"
    });
    bank.push({
      externalId: `B-${id1}`, amount, currency: "INR", transactionDate: dateStr, referenceId: `PAY-DUP-${i}`, description: "ACH SETTLEMENT ACME"
    });

    addMatch(`P-${id1}`, `S-${id1}`);
    addMatch(`S-${id1}`, `B-${id1}`);
  }

  // 7. Truly unmatched transactions (10)
  for (let i = 0; i < 5; i++) {
    const id = nextId++;
    const amount = (random() * 10000 + 100).toFixed(2);
    const dateStr = baseDate.toISOString().split("T")[0];

    // Orphan payment
    payments.push({
      externalId: `P-${id}`, amount, currency: "INR", transactionDate: dateStr, referenceId: `PAY-ORPHAN-${id}`, merchantName: "Lost Corp"
    });
    
    // Orphan bank (e.g. unknown fee)
    const id2 = nextId++;
    bank.push({
      externalId: `B-${id2}`, amount: "25.00", currency: "INR", transactionDate: dateStr, referenceId: `FEE-${id2}`, description: "MONTHLY MAINTENANCE FEE"
    });
  }

  // 8. Medium-confidence fuzzy matches (5)
  for (let i = 0; i < 5; i++) {
    const id = nextId++;
    const ref = `PAY-${1000 + id}`;
    const amount = (random() * 10000 + 100).toFixed(2);
    const dateStr = baseDate.toISOString().split("T")[0];

    // Missing ref + wrong date + messy desc
    payments.push({
      externalId: `P-${id}`, amount, currency: "INR", transactionDate: dateStr, referenceId: ref, merchantName: "Acme Ltd"
    });
    
    const offsetDate = new Date(baseDate); offsetDate.setDate(offsetDate.getDate() + 5);
    
    settlements.push({
      externalId: `S-${id}`, amount, currency: "INR", transactionDate: offsetDate.toISOString().split("T")[0], referenceId: "", description: "A.C.M.E SETT."
    });
    
    addMatch(`P-${id}`, `S-${id}`);
  }

  // 9. Settlement timing differences (weekend delays) (5)
  for (let i = 0; i < 5; i++) {
    const id = nextId++;
    const ref = `PAY-${1000 + id}`;
    const amount = (random() * 10000 + 100).toFixed(2);
    const friDate = new Date("2026-08-21"); // Friday
    const monDate = new Date("2026-08-24"); // Monday

    payments.push({
      externalId: `P-${id}`, amount, currency: "INR", transactionDate: friDate.toISOString().split("T")[0], referenceId: ref, merchantName: "Weekend Corp"
    });
    settlements.push({
      externalId: `S-${id}`, amount, currency: "INR", transactionDate: friDate.toISOString().split("T")[0], referenceId: ref, description: "WEEKEND CORP SETTLEMENT"
    });
    bank.push({
      externalId: `B-${id}`, amount, currency: "INR", transactionDate: monDate.toISOString().split("T")[0], referenceId: ref, description: "ACH SETTLEMENT WEEKEND CORP"
    });

    addMatch(`P-${id}`, `S-${id}`);
    addMatch(`S-${id}`, `B-${id}`);
  }

  // 10. Records contributing to cash variance (missing ledger entries for bank deposits) (5)
  for (let i = 0; i < 5; i++) {
    const id = nextId++;
    const amount = (random() * 10000 + 100).toFixed(2);
    const dateStr = baseDate.toISOString().split("T")[0];

    // Bank deposit that has no corresponding ledger entry
    bank.push({
      externalId: `B-${id}`, amount, currency: "INR", transactionDate: dateStr, referenceId: `UNEXPECTED-${id}`, description: "WIRE TRANSFER FROM UNKNOWN"
    });
  }

  const dataDir = path.join(process.cwd(), "data");
  if (!fs.existsSync(dataDir)) {
    fs.mkdirSync(dataDir);
  }

  function toCsv(arr: any[]) {
    if (arr.length === 0) return "";
    const headers = Object.keys(arr[0]).join(",");
    const rows = arr.map(obj => Object.values(obj).map(v => typeof v === 'string' && v.includes(',') ? `"${v}"` : v).join(",")).join("\n");
    return `${headers}\n${rows}`;
  }

  fs.writeFileSync(path.join(dataDir, "payments.csv"), toCsv(payments));
  fs.writeFileSync(path.join(dataDir, "settlements.csv"), toCsv(settlements));
  fs.writeFileSync(path.join(dataDir, "bank.csv"), toCsv(bank));
  fs.writeFileSync(path.join(dataDir, "ledger.csv"), toCsv(ledgers));
  fs.writeFileSync(path.join(dataDir, "ground_truth.csv"), toCsv(groundTruth));

  const total = payments.length + settlements.length + bank.length + ledgers.length;
  console.log(`Synthetic data generated deterministically in /data folder. Total records: ${total}`);
}

generate();

File: scripts/generate_synthetic_data.py
import pandas as pd
import random
import os
from datetime import datetime, timedelta
from faker import Faker

# Use Faker for realistic names and references
fake = Faker()
Faker.seed(42)
random.seed(42)

def generate_synthetic_data():
    base_date = datetime.strptime("2026-08-18", "%Y-%m-%d")
    
    payments = []
    settlements = []
    ledgers = []
    ground_truth = []
    
    next_id = 1
    
    # Helper to format amounts properly
    def format_amount(amount):
        return f"{amount:.2f}"
    
    # 1. Exact Matches (60 records)
    for _ in range(60):
        t_id = next_id
        next_id += 1
        ref = f"PAY-{1000 + t_id}"
        amount = random.uniform(100.0, 10000.0)
        
        merchant = fake.company()
        customer = merchant + " Inc."
        
        payments.append({
            "externalId": f"P-{t_id}",
            "amount": format_amount(amount),
            "currency": "INR",
            "transactionDate": base_date.strftime("%Y-%m-%d"),
            "referenceId": ref,
            "merchantName": merchant
        })
        
        settlements.append({
            "externalId": f"S-{t_id}",
            "amount": format_amount(amount),
            "currency": "INR",
            "transactionDate": base_date.strftime("%Y-%m-%d"),
            "referenceId": ref,
            "description": f"{merchant.upper()} SETTLEMENT"
        })
        
        ledgers.append({
            "externalId": f"L-{t_id}",
            "amount": format_amount(amount),
            "currency": "INR",
            "transactionDate": base_date.strftime("%Y-%m-%d"),
            "referenceId": f"INV-{900 + t_id}",
            "customerName": customer
        })
        
        ground_truth.append({"source_transaction": f"P-{t_id}", "target_transaction": f"S-{t_id}", "is_match": "true"})
        ground_truth.append({"source_transaction": f"S-{t_id}", "target_transaction": f"L-{t_id}", "is_match": "true"})

    # 2. Date Offsets (15 records) - Bank settlement happens 1-2 days later
    for _ in range(15):
        t_id = next_id
        next_id += 1
        ref = f"PAY-{1000 + t_id}"
        amount = random.uniform(100.0, 10000.0)
        merchant = fake.company()
        
        offset_days = random.randint(1, 2)
        offset_date = base_date + timedelta(days=offset_days)
        
        payments.append({
            "externalId": f"P-{t_id}",
            "amount": format_amount(amount),
            "currency": "INR",
            "transactionDate": base_date.strftime("%Y-%m-%d"),
            "referenceId": ref,
            "merchantName": merchant
        })
        
        settlements.append({
            "externalId": f"S-{t_id}",
            "amount": format_amount(amount),
            "currency": "INR",
            "transactionDate": offset_date.strftime("%Y-%m-%d"),
            "referenceId": ref,
            "description": f"{merchant.upper()} SETTLEMENT"
        })
        
        ledgers.append({
            "externalId": f"L-{t_id}",
            "amount": format_amount(amount),
            "currency": "INR",
            "transactionDate": base_date.strftime("%Y-%m-%d"),
            "referenceId": f"INV-{900 + t_id}",
            "customerName": merchant + " Corp"
        })
        
        ground_truth.append({"source_transaction": f"P-{t_id}", "target_transaction": f"S-{t_id}", "is_match": "true"})
        ground_truth.append({"source_transaction": f"S-{t_id}", "target_transaction": f"L-{t_id}", "is_match": "true"})

    # 3. Description mismatches (8 records) - Messy string variations
    for _ in range(8):
        t_id = next_id
        next_id += 1
        ref = f"PAY-{1000 + t_id}"
        amount = random.uniform(100.0, 10000.0)
        merchant_base = fake.company()
        
        payments.append({
            "externalId": f"P-{t_id}",
            "amount": format_amount(amount),
            "currency": "INR",
            "transactionDate": base_date.strftime("%Y-%m-%d"),
            "referenceId": ref,
            "merchantName": merchant_base
        })
        
        # Strip vowels and add noise for messy description
        messy_desc = merchant_base.upper().replace('A', '').replace('E', '') + " STTLMNT"
        
        settlements.append({
            "externalId": f"S-{t_id}",
            "amount": format_amount(amount),
            "currency": "INR",
            "transactionDate": base_date.strftime("%Y-%m-%d"),
            "referenceId": ref,
            "description": messy_desc
        })
        
        ledgers.append({
            "externalId": f"L-{t_id}",
            "amount": format_amount(amount),
            "currency": "INR",
            "transactionDate": base_date.strftime("%Y-%m-%d"),
            "referenceId": f"INV-{900 + t_id}",
            "customerName": merchant_base + " LLC"
        })
        
        ground_truth.append({"source_transaction": f"P-{t_id}", "target_transaction": f"S-{t_id}", "is_match": "true"})
        ground_truth.append({"source_transaction": f"S-{t_id}", "target_transaction": f"L-{t_id}", "is_match": "true"})

    # 4. Amount mismatches / fees (5 records) - Minor variations (< 1%)
    for _ in range(5):
        t_id = next_id
        next_id += 1
        ref = f"PAY-{1000 + t_id}"
        base_amount = random.uniform(500.0, 10000.0)
        merchant = fake.company()
        
        payments.append({
            "externalId": f"P-{t_id}",
            "amount": format_amount(base_amount),
            "currency": "INR",
            "transactionDate": base_date.strftime("%Y-%m-%d"),
            "referenceId": ref,
            "merchantName": merchant
        })
        
        # Subtract a small fee for the bank settlement
        fee = random.uniform(1.0, 10.0)
        
        settlements.append({
            "externalId": f"S-{t_id}",
            "amount": format_amount(base_amount - fee),
            "currency": "INR",
            "transactionDate": base_date.strftime("%Y-%m-%d"),
            "referenceId": ref,
            "description": f"{merchant.upper()} SETTLEMENT (NET)"
        })
        
        ledgers.append({
            "externalId": f"L-{t_id}",
            "amount": format_amount(base_amount),
            "currency": "INR",
            "transactionDate": base_date.strftime("%Y-%m-%d"),
            "referenceId": f"INV-{900 + t_id}",
            "customerName": merchant
        })
        
        # This will test fuzzy logic based on tolerance
        ground_truth.append({"source_transaction": f"P-{t_id}", "target_transaction": f"S-{t_id}", "is_match": "true"})
        ground_truth.append({"source_transaction": f"S-{t_id}", "target_transaction": f"L-{t_id}", "is_match": "true"})

    # 5. Duplicates (4 records) - Same exact payment happens twice
    for _ in range(4):
        t_id = next_id
        next_id += 1
        ref = f"PAY-{1000 + t_id}"
        amount = format_amount(random.uniform(100.0, 10000.0))
        merchant = fake.company()
        
        payment = {
            "externalId": f"P-{t_id}",
            "amount": amount,
            "currency": "INR",
            "transactionDate": base_date.strftime("%Y-%m-%d"),
            "referenceId": ref,
            "merchantName": merchant
        }
        
        payments.append(payment)
        payments.append(payment.copy()) # Duplicate!
        
        settlements.append({
            "externalId": f"S-{t_id}",
            "amount": amount,
            "currency": "INR",
            "transactionDate": base_date.strftime("%Y-%m-%d"),
            "referenceId": ref,
            "description": f"{merchant.upper()} SETTLEMENT"
        })
        
        ledgers.append({
            "externalId": f"L-{t_id}",
            "amount": amount,
            "currency": "INR",
            "transactionDate": base_date.strftime("%Y-%m-%d"),
            "referenceId": f"INV-{900 + t_id}",
            "customerName": merchant
        })
        
        ground_truth.append({"source_transaction": f"P-{t_id}", "target_transaction": f"S-{t_id}", "is_match": "true"})

    # 6. Missing ledger entries (5 records) - Payment hit bank, but not recorded in Ledger
    for _ in range(5):
        t_id = next_id
        next_id += 1
        ref = f"PAY-{1000 + t_id}"
        amount = format_amount(random.uniform(100.0, 10000.0))
        merchant = fake.company()
        
        payments.append({
            "externalId": f"P-{t_id}",
            "amount": amount,
            "currency": "INR",
            "transactionDate": base_date.strftime("%Y-%m-%d"),
            "referenceId": ref,
            "merchantName": merchant
        })
        
        settlements.append({
            "externalId": f"S-{t_id}",
            "amount": amount,
            "currency": "INR",
            "transactionDate": base_date.strftime("%Y-%m-%d"),
            "referenceId": ref,
            "description": f"{merchant.upper()} SETTLEMENT"
        })
        
        ground_truth.append({"source_transaction": f"P-{t_id}", "target_transaction": f"S-{t_id}", "is_match": "true"})
        # No ledger appended.

    # 7. Ambiguous candidates (3 records) - Multiple settlements for same exact amount on same day
    for _ in range(3):
        t_id = next_id
        next_id += 1
        ref = f"PAY-{1000 + t_id}"
        amount = "5000.00" # Explicit hardcoded collision
        merchant = fake.company()
        
        payments.append({
            "externalId": f"P-{t_id}",
            "amount": amount,
            "currency": "INR",
            "transactionDate": base_date.strftime("%Y-%m-%d"),
            "referenceId": ref,
            "merchantName": merchant
        })
        
        settlements.append({
            "externalId": f"S-{t_id}-A",
            "amount": amount,
            "currency": "INR",
            "transactionDate": base_date.strftime("%Y-%m-%d"),
            "referenceId": f"REF-{fake.bban()}", 
            "description": f"{merchant.upper()} BATCH 1"
        })
        
        settlements.append({
            "externalId": f"S-{t_id}-B",
            "amount": amount,
            "currency": "INR",
            "transactionDate": base_date.strftime("%Y-%m-%d"),
            "referenceId": f"REF-{fake.bban()}",
            "description": f"{merchant.upper()} BATCH 2"
        })
        
        # Only one is the real match
        ground_truth.append({"source_transaction": f"P-{t_id}", "target_transaction": f"S-{t_id}-A", "is_match": "true"})

    # Create data directory
    data_dir = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "data")
    os.makedirs(data_dir, exist_ok=True)
    
    # Save to CSV using Pandas
    pd.DataFrame(payments).to_csv(os.path.join(data_dir, "payments.csv"), index=False)
    pd.DataFrame(settlements).to_csv(os.path.join(data_dir, "settlements.csv"), index=False)
    pd.DataFrame(ledgers).to_csv(os.path.join(data_dir, "ledger.csv"), index=False)
    pd.DataFrame(ground_truth).to_csv(os.path.join(data_dir, "ground_truth.csv"), index=False)

    print(f"Synthetic data generated successfully in {data_dir}")
    print(f"Total Payments generated: {len(payments)}")

if __name__ == "__main__":
    generate_synthetic_data()
